src/modules/rlm_python/rlm_python.c

```diff
@@ -488,95 +488,147 @@ static int mod_populate_vptuple(rlm_python_t const *inst, request_t *request, Py
 	fr_cursor_t	cursor;
-	fr_pair_t	*vp;
-	PyObject	*p_ret = NULL;
-	PyObject	*p_arg = NULL;
-	int		tuple_len;
-	rlm_rcode_t	rcode = RLM_MODULE_OK;
+	fr_pair_t 	*vp;
+
+	/* If vps is NULL, return None */
+	if (vps == NULL) {
+		Py_INCREF(Py_None);
+		PyTuple_SET_ITEM(p_arg, pos, Py_None);
+		return true;
+	}
 
 	/*
 	 *	We will pass a tuple containing (name, value) tuples
 	 *	We can safely use the Python function to build up a
 	 *	tuple, since the tuple is not used elsewhere.
 	 *
-	 *	Determine the size of our tuple by walking through the packet.
-	 *	If request is NULL, pass None.
+	 *	Determine the size of our tuple by walking through the vps.
 	 */
-	tuple_len = 0;
-	if (request != NULL) {
-		for (vp = fr_cursor_init(&cursor, &request->request_pairs);
-		     vp;
-		     vp = fr_cursor_next(&cursor)) tuple_len++;
-	}
+	for (vp = fr_cursor_init(&cursor, &vps); vp; vp = fr_cursor_next(&cursor)) tuplelen++;
 
-	if (tuple_len == 0) {
-		Py_INCREF(Py_None);
-		p_arg = Py_None;
-	} else {
-		int i = 0;
-		if ((p_arg = PyTuple_New(tuple_len)) == NULL) {
-			rcode = RLM_MODULE_FAIL;
-			goto finish;
+	if ((vps_tuple = PyTuple_New(tuplelen)) == NULL) goto error;
+
+	for (vp = fr_cursor_init(&cursor, &vps); vp; vp = fr_cursor_next(&cursor), i++) {
+		PyObject *pPair = NULL;
+
+		/* The inside tuple has two only: */
+		if ((pPair = PyTuple_New(2)) == NULL) goto error;
+
+		if (mod_populate_vptuple(inst, request, pPair, vp) == 0) {
+			/* Put the tuple inside the container */
+			PyTuple_SET_ITEM(vps_tuple, i, pPair);
+		} else {
+			Py_DECREF(pPair);
+			goto error;
 		}
+	}
+	PyTuple_SET_ITEM(p_arg, pos, vps_tuple);
+	return true;
 
-		for (vp = fr_cursor_init(&cursor, &request->request_pairs);
-		     vp;
-		     vp = fr_cursor_next(&cursor), i++) {
-			PyObject *pp;
+error:
+	Py_XDECREF(vps_tuple);
+	return false;
+}
 
-			/* The inside tuple has two only: */
-			if ((pp = PyTuple_New(2)) == NULL) {
-				rcode = RLM_MODULE_FAIL;
-				goto finish;
-			}
+static unlang_action_t do_python_single(rlm_rcode_t *p_result,
+					rlm_python_t const *inst, request_t *request, PyObject *p_func, char const *funcname)
+{
+	PyObject	*p_ret = NULL;
+	PyObject	*p_arg = NULL;
+	PyObject 	*p_dict = NULL;
+	PyObject    *p_tuple_int = NULL;
+	PyObject    *update_dict = NULL;
+	rlm_rcode_t	rcode;
 
-			if (mod_populate_vptuple(inst, request, pp, vp) == 0) {
-				/* Put the tuple inside the container */
-				PyTuple_SET_ITEM(p_arg, i, pp);
-			} else {
-				Py_INCREF(Py_None);
-				PyTuple_SET_ITEM(p_arg, i, Py_None);
-				Py_DECREF(pp);
-			}
+	/* Default return value is "OK, continue" */
+	rcode = RLM_MODULE_OK;
+
+	/*
+	 * p_arg is a 4-tuple with (Request, Reply, Control and State)
+	 * If some list is not available, NONE is used instead
+	 */
+	p_arg = PyTuple_New(4);
+	if (!p_arg) {
+		ERROR("%s - Memory cannot be allocated for PyTyple_New", funcname);
+		rcode = RLM_MODULE_FAIL;
+		goto finish;
+	}
+
+	/* If there is a request, fill in the first 4 attribute lists */
```

Comment:
there should always be a request?  If the function is being called from `recv foo` or xlat, there's always a request.